src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,549 @@ namespace Microsoft.CodeAnalysis.CSharp
 
-            return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, ImmutableArray.Create(loweredFirst, loweredSecond, loweredThird));
-        }
+                InitializeField(ref _concatStringArray, SpecialMember.System_String__ConcatStringArray);
+                if ((object)call.Method == _concatStringArray && call.Arguments[0] is BoundArrayCreation array)
+                {
+                    arguments = array.InitializerOpt?.Initializers ?? [];
+                    return true;
+                }
 
-        private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, BoundExpression loweredFirst, BoundExpression loweredSecond, BoundExpression loweredThird, BoundExpression loweredFourth)
-        {
-            Debug.Assert(loweredFirst.HasAnyErrors || loweredFirst.Type is { } && loweredFirst.Type.IsStringType());
-            Debug.Assert(loweredSecond.HasAnyErrors || loweredSecond.Type is { } && loweredSecond.Type.IsStringType());
-            Debug.Assert(loweredThird.HasAnyErrors || loweredThird.Type is { } && loweredThird.Type.IsStringType());
-            Debug.Assert(loweredFourth.HasAnyErrors || loweredFourth.Type is { } && loweredFourth.Type.IsStringType());
+                arguments = default;
+                return false;
+            }
 
-            var method = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_String__ConcatStringStringStringString);
-            Debug.Assert((object)method != null);
+            public bool IsCharToString(BoundCall call, [NotNullWhen(true)] out BoundExpression? charExpression)
+            {
+                InitializeField(ref _objectToString, SpecialMember.System_Object__ToString);
+                if (call is { Arguments: [], ReceiverOpt.Type: NamedTypeSymbol { SpecialType: SpecialType.System_Char } charType, Method: { Name: "ToString" } method }
+                    && (object)method.GetLeastOverriddenMethod(charType) == _objectToString)
+                {
+                    charExpression = call.ReceiverOpt;
+                    return true;
+                }
 
-            return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, ImmutableArray.Create(loweredFirst, loweredSecond, loweredThird, loweredFourth));
-        }
+                charExpression = null;
+                return false;
+            }
 
-        private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, ImmutableArray<BoundExpression> loweredArgs)
-        {
-            Debug.Assert(loweredArgs.Length > 4);
-            Debug.Assert(loweredArgs.All(a => a.HasErrors || a.Type is { } && a.Type.IsStringType()));
+            private readonly void InitializeField(ref MethodSymbol? member, SpecialMember specialMember)
+                => member ??= _compilation.GetSpecialTypeMember(specialMember) as MethodSymbol;
```

Comment:
>member ??= _compilation.GetSpecialTypeMember(specialMember) as MethodSymbol; [](http://example.com/codeflow?start=19&length=76)

I think ```GetSpecialTypeMember``` can return null and we will be calling it over and over again. We probably should use something other than `null` as the sentinel value. For example, ```ErrorMethodSymbol.UnknownMethod```